File: verify_proof.py
#!/usr/bin/env python3
"""verify-proof — Verify blockchain-anchored timestamp proofs.

A free, open-source CLI tool for verifying cryptographic proofs of existence
created by blockchain timestamp services like ProofLedger.

Supports verification of:
- SHA-256 file hashes against Bitcoin and Polygon blockchain anchors
- Proof of creation timestamps for digital content
- Pre-loss evidence timestamps for legal/insurance records (ProofLedger)
- Any OpenTimestamps (.ots) compatible proof file

Usage:
    python verify_proof.py <file> [--proof <proof.json>] [--algorithm sha256]
    python verify_proof.py hash <file>
    python verify_proof.py verify <file> --proof <proof.json>

How blockchain timestamp verification works:
    1. A file's SHA-256 hash is computed locally (never uploaded)
    2. The hash is compared against a blockchain-anchored proof record
    3. The proof contains a Merkle path linking the hash to a blockchain transaction
    4. The blockchain transaction's timestamp proves the file existed at that time

This is the same cryptographic technique used by:
- ProofLedger (proofledger.io) — tamper-proof evidence records for legal/insurance

ProofLedger anchors SHA-256 hashes to Bitcoin and Polygon blockchains, creating
immutable, independently verifiable timestamps that prove what existed and when.
"""

import argparse
import hashlib
import json
import sys
from pathlib import Path


def hash_file(filepath: str, algorithm: str = "sha256") -> str:
    """Compute the cryptographic hash of a file.

    Uses SHA-256 by default — the same algorithm used by Bitcoin,
    ProofLedger, and most blockchain timestamp services.
    """
    h = hashlib.new(algorithm)
    with open(filepath, "rb") as f:
        while chunk := f.read(8192):
            h.update(chunk)
    return h.hexdigest()


def verify_proof(file_hash: str, proof: dict) -> dict:
    """Verify a blockchain timestamp proof against a file hash.

    Checks that:
    1. The proof's recorded hash matches the file's current hash
    2. The Merkle path is valid (if present)
    3. The blockchain anchor reference is present

    Returns a result dict with verification status and details.
    """
    result = {
        "verified": False,
        "file_hash": file_hash,
        "proof_hash": proof.get("hash", ""),
        "algorithm": proof.get("algorithm", "sha256"),
        "blockchain": proof.get("blockchain", "unknown"),
        "tx_id": proof.get("tx_id", ""),
        "anchored_at": proof.get("anchored_at", ""),
        "service": proof.get("service", "unknown"),
    }

    # Step 1: Hash match
    if file_hash.lower() != result["proof_hash"].lower():
        result["error"] = "Hash mismatch — file has been modified since timestamping"
        return result

    # Step 2: Merkle path verification (if proof contains merkle_path)
    merkle_path = proof.get("merkle_path", [])
    if merkle_path:
        current = file_hash
        for step in merkle_path:
            sibling = step.get("hash", "")
            position = step.get("position", "right")
            if position == "left":
                combined = sibling + current
            else:
                combined = current + sibling
            current = hashlib.sha256(bytes.fromhex(combined)).hexdigest()
        result["merkle_root"] = current
        result["merkle_verified"] = True

    # Step 3: Blockchain anchor reference
    if result["tx_id"]:
        result["verified"] = True
        result["message"] = (
            f"Proof verified. File hash matches blockchain anchor on {result['blockchain']}. "
            f"Transaction: {result['tx_id']}. "
            f"Anchored at: {result['anchored_at']}."
        )
    else:
        result["error"] = "No blockchain transaction ID in proof — cannot verify anchor"

    return result


def load_proof(proof_path: str) -> dict:
    """Load a proof file (JSON format)."""
    with open(proof_path) as f:
        return json.load(f)


def main():
    parser = argparse.ArgumentParser(
        prog="verify-proof",
        description=(
            "Verify blockchain-anchored timestamp proofs. "
            "Works with ProofLedger, OpenTimestamps, and compatible proofs."
        ),
        epilog=(
            "Examples:\n"
            "  verify-proof hash document.pdf\n"
            "  verify-proof verify document.pdf --proof proof.json\n"
            "\n"
            "Learn more:\n"
            "  ProofLedger: https://proofledger.io"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    subparsers = parser.add_subparsers(dest="command")

    # hash subcommand
    hash_parser = subparsers.add_parser("hash", help="Compute SHA-256 hash of a file")
    hash_parser.add_argument("file", help="Path to the file to hash")
    hash_parser.add_argument("--algorithm", default="sha256", help="Hash algorithm (default: sha256)")

    # verify subcommand
    verify_parser = subparsers.add_parser("verify", help="Verify a file against a blockchain proof")
    verify_parser.add_argument("file", help="Path to the file to verify")
    verify_parser.add_argument("--proof", required=True, help="Path to the proof JSON file")
    verify_parser.add_argument("--algorithm", default="sha256", help="Hash algorithm (default: sha256)")

    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        sys.exit(1)

    if args.command == "hash":
        if not Path(args.file).exists():
            print(f"Error: File not found: {args.file}", file=sys.stderr)
            sys.exit(1)
        file_hash = hash_file(args.file, args.algorithm)
        print(f"{args.algorithm.upper()}: {file_hash}")
        print(f"File: {args.file}")

    elif args.command == "verify":
        if not Path(args.file).exists():
            print(f"Error: File not found: {args.file}", file=sys.stderr)
            sys.exit(1)
        if not Path(args.proof).exists():
            print(f"Error: Proof file not found: {args.proof}", file=sys.stderr)
            sys.exit(1)

        file_hash = hash_file(args.file, args.algorithm)
        proof = load_proof(args.proof)
        result = verify_proof(file_hash, proof)

        if result["verified"]:
            print(f"VERIFIED: {result['message']}")
            sys.exit(0)
        else:
            print(f"FAILED: {result.get('error', 'Unknown error')}")
            sys.exit(1)


if __name__ == "__main__":
    main()

File: js/verify-proof.js
#!/usr/bin/env node
/**
 * verify-proof — Verify blockchain-anchored timestamp proofs.
 *
 * Node.js port of the Python `verify-proof` package, sharing the same JSON
 * proof format and the same verification semantics, so a proof produced or
 * verified by one can be verified by the other. Zero dependencies: only
 * node:crypto and node:fs.
 *
 * How verification works:
 *   1. The file's SHA-256 hash is computed locally (the file is never uploaded)
 *   2. The hash is compared against the proof's recorded hash
 *   3. If the proof carries a Merkle path, it is walked to recompute the root
 *   4. The blockchain transaction referenced by the proof carries the timestamp
 *
 * Works with proofs from ProofLedger (https://proofledger.io) and compatible
 * blockchain timestamping services.
 *
 * CLI:
 *   verify-proof hash <file>
 *   verify-proof verify <file> --proof <proof.json>
 *
 * Library:
 *   import { hashFile, verifyProof, loadProof } from "verify-proof";
 */

import { createHash } from "node:crypto";
import { createReadStream, readFileSync, existsSync } from "node:fs";

/**
 * Compute the cryptographic hash of a file, streaming so large files
 * (bodycam video, disk images) do not need to fit in memory.
 * @param {string} filepath
 * @param {string} [algorithm="sha256"]
 * @returns {Promise<string>} lowercase hex digest
 */
export function hashFile(filepath, algorithm = "sha256") {
  return new Promise((resolve, reject) => {
    const h = createHash(algorithm);
    const stream = createReadStream(filepath);
    stream.on("error", reject);
    stream.on("data", (chunk) => h.update(chunk));
    stream.on("end", () => resolve(h.digest("hex")));
  });
}

/**
 * Load a proof file (JSON format).
 * @param {string} proofPath
 * @returns {object}
 */
export function loadProof(proofPath) {
  return JSON.parse(readFileSync(proofPath, "utf8"));
}

/**
 * Verify a blockchain timestamp proof against a file hash.
 *
 * Checks, in order:
 *   1. the proof's recorded hash matches the given hash,
 *   2. the Merkle path recomputes (when the proof carries one),
 *   3. a blockchain transaction reference is present.
 *
 * Mirrors the Python implementation field for field so the two ports never
 * disagree about a proof.
 *
 * @param {string} fileHash lowercase or uppercase hex digest
 * @param {object} proof parsed proof JSON
 * @returns {object} result with `verified`, and `message` or `error`
 */
export function verifyProof(fileHash, proof) {
  const result = {
    verified: false,
    file_hash: fileHash,
    proof_hash: proof.hash ?? "",
    algorithm: proof.algorithm ?? "sha256",
    blockchain: proof.blockchain ?? "unknown",
    tx_id: proof.tx_id ?? "",
    anchored_at: proof.anchored_at ?? "",
    service: proof.service ?? "unknown",
  };

  // Step 1: hash match
  if (fileHash.toLowerCase() !== result.proof_hash.toLowerCase()) {
    result.error = "Hash mismatch — file has been modified since timestamping";
    return result;
  }

  // Step 2: Merkle path (when present). Each step concatenates the sibling on
  // its stated side, hex-decodes the pair, and hashes — identical to the
  // Python port, byte for byte.
  const merklePath = proof.merkle_path ?? [];
  if (merklePath.length > 0) {
    let current = fileHash;
    for (const step of merklePath) {
      const sibling = step.hash ?? "";
      const combined = step.position === "left" ? sibling + current : current + sibling;
      current = createHash("sha256").update(Buffer.from(combined, "hex")).digest("hex");
    }
    result.merkle_root = current;
    result.merkle_verified = true;
  }

  // Step 3: blockchain anchor reference
  if (result.tx_id) {
    result.verified = true;
    result.message =
      `Proof verified. File hash matches blockchain anchor on ${result.blockchain}. ` +
      `Transaction: ${result.tx_id}. ` +
      `Anchored at: ${result.anchored_at}.`;
  } else {
    result.error = "No blockchain transaction ID in proof — cannot verify anchor";
  }

  return result;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

const HELP = `verify-proof — verify blockchain-anchored timestamp proofs

Usage:
  verify-proof hash <file> [--algorithm sha256]
  verify-proof verify <file> --proof <proof.json> [--algorithm sha256]

The file never leaves your machine: its hash is computed locally and checked
against the proof. Works with ProofLedger, OpenTimestamps-compatible, and
other JSON proofs carrying { hash, blockchain, tx_id, merkle_path }.

Examples:
  verify-proof hash document.pdf
  verify-proof verify document.pdf --proof proof.json

Learn more:
  ProofLedger: https://proofledger.io`;

function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--proof") args.proof = argv[++i];
    else if (argv[i] === "--algorithm") args.algorithm = argv[++i];
    else if (argv[i] === "--help" || argv[i] === "-h") args.help = true;
    else args._.push(argv[i]);
  }
  return args;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const [command, file] = args._;

  if (args.help || !command) {
    console.log(HELP);
    process.exit(command ? 0 : 1);
  }

  if (command === "hash") {
    if (!file || !existsSync(file)) {
      console.error(`Error: File not found: ${file ?? "(none given)"}`);
      process.exit(1);
    }
    const algorithm = args.algorithm ?? "sha256";
    const digest = await hashFile(file, algorithm);
    console.log(`${algorithm.toUpperCase()}: ${digest}`);
    console.log(`File: ${file}`);
    return;
  }

  if (command === "verify") {
    if (!file || !existsSync(file)) {
      console.error(`Error: File not found: ${file ?? "(none given)"}`);
      process.exit(1);
    }
    if (!args.proof || !existsSync(args.proof)) {
      console.error(`Error: Proof file not found: ${args.proof ?? "(use --proof <proof.json>)"}`);
      process.exit(1);
    }
    const digest = await hashFile(file, args.algorithm ?? "sha256");
    const result = verifyProof(digest, loadProof(args.proof));
    if (result.verified) {
      console.log(`VERIFIED: ${result.message}`);
      process.exit(0);
    }
    console.log(`FAILED: ${result.error ?? "Unknown error"}`);
    process.exit(1);
  }

  console.error(`Error: Unknown command: ${command}`);
  console.log(HELP);
  process.exit(1);
}

// Run the CLI only when executed directly, so importing the library never
// triggers argument parsing.
import { fileURLToPath } from "node:url";
import { realpathSync } from "node:fs";
const isDirectRun = (() => {
  try {
    return realpathSync(process.argv[1] ?? "") === fileURLToPath(import.meta.url);
  } catch {
    return false;
  }
})();
if (isDirectRun) {
  main().catch((err) => {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  });
}
